pkg/clusterid/parse_cluster_id.go

```diff
@@ -5,13 +5,59 @@ import (
 
 	"github.com/pkg/errors"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/buildinfo"
+	"github.com/stackrox/rox/pkg/logging"
 	"github.com/stackrox/rox/pkg/mtls"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
+var (
+	once     sync.Once
+	instance *parserWrapper
+	log      = logging.LoggerForModule()
+)
+
+// Parser defines an interface with the function to parse the cluster ID from the service cert.
+type Parser interface {
+	ParseClusterIDFromServiceCert(expectedServiceType storage.ServiceType) (string, error)
+}
+
+// parserWrapper is a singleton that wraps the Parser interface.
+// This allows us to override the implementation of the Parser interface for testing purposes.
+// This override mechanism is protected in ReleaseBuilds, and it will only work for DevelopmentBuilds.
+type parserWrapper struct {
+	parser Parser
+}
+
+// parserImpl is the implementation of the Parser interface.
+type parserImpl struct {
+}
+
+// GetParser returns the parserWrapper singleton.
+func GetParser() *parserWrapper {
+	once.Do(func() {
+		instance = &parserWrapper{
+			parser: &parserImpl{},
+		}
+	})
+	return instance
+}
+
+// Override the internal parser. This should only be used for testing.
+func (p *parserWrapper) Override(parser Parser) {
+	if buildinfo.ReleaseBuild {
+		// This is not allowed in release builds
+		log.Errorf("ParseClusterIDFromServiceCert should not be override in production")
+		return
+	}
+	p.parser = parser
```

Comment:
Maybe we should put this behind some lock? I assume there's a race if `GetParser` is called together with `Override`. You can use a `RWMutex` since `p` is read much more often than its written.

Reply:
I don’t think you can have a race between `GetParser` and `Override` if we keep the wrapper since `GetParser` needs to be called always before `Override`. I saw a data race between multiple calls of `Override` so I added a mutex only in the `Override` function.

Reply:
See comment above about race.

Reply:
Makes sense 👍